### PR TITLE
Task/34 - forgot password

### PR DESCRIPTION
**Pull request by Zalezny**
image

Check send email result:

https://user-images.githubusercontent.com/65240240/231847449-190c74f1-d840-40d2-abaa-8025000e38a1.mp4

### CODE REVIEW

File: app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+
+        val etEmail = binding.eTextEmailAddress
+        val requestBtn = binding.requestButton
+
+        EmailValidator.attach(etEmail)
+
+        etEmail.addTextChangedListener {
+            forgotPasswordViewModel.setError(etEmail.error?.toString() ?: "")
```
Comment (msarelo) [inline app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt:37]:
```suggestion
            forgotPasswordViewModel.setError(etEmail.error?.toString().orEmpty())
```

File: app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        forgotPasswordViewModel.error.observe(viewLifecycleOwner) { errorMsg ->
+            if (errorMsg.isNullOrEmpty().not()) {
+                requestBtn.backgroundTintList = getColor(R.color.p_30)
+            } else {
+                requestBtn.backgroundTintList = getColor(R.color.p_60)
+            }
```
Comment (msarelo) [inline app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt:45]:
```suggestion
            if (errorMsg.isNullOrEmpty()) {
                requestBtn.backgroundTintList = getColor(R.color.p_60)
            } else {
                requestBtn.backgroundTintList = getColor(R.color.p_30)
            }
```

File: app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        forgotPasswordViewModel.isSuccess.observe(viewLifecycleOwner) { isSuccess ->
+            if (isSuccess) {
+                TODO("Success of sent reset Password on email")
```
Comment (msarelo) [inline app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordFragment.kt:49]: Please add Snackbar as at this screen:
https://zpl.io/XYo0LxK

File: app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    fun resetPassword(email: String) {
+        viewModelScope.launch {
+
+            resetPasswordUseCase.invoke(ResetPasswordUseCase.ResetParam(email)).collect { result ->
+                if (result.isSuccess) {
+                    _isSuccess.postValue(true)
+                } else {
+                    _isSuccess.postValue(false)
+                }
```
Comment (msarelo) [inline app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/ForgotPasswordViewModel.kt:30]:
```suggestion
                    _isSuccess.postValue(result.isSuccess)
                }
```

File: app/src/main/res/layout/fragment_forgot_password.xml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/size_0.25x"
+        android:layout_marginStart="10dp"
+        android:text="@string/forgot_password_text"
+        android:textColor="@color/n_120"
+        android:textSize="14sp"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/tv_forgot_pass" />
+
+
```
Comment (msarelo) [inline app/src/main/res/layout/fragment_forgot_password.xml:64]:
```suggestion
```

File: app/src/main/res/layout/fragment_forgot_password.xml [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        android:id="@+id/email_address"
+        style="@style/Text2"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/size_3x"
+        android:text="@string/e_mail"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/tv_forgot_pass_text" />
+
+
```
Comment (msarelo) [inline app/src/main/res/layout/fragment_forgot_password.xml:75]:
```suggestion
```

File: app/src/main/res/layout/fragment_forgot_password.xml [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            android:text="@string/sign_in_now"
+            app:layout_constraintEnd_toEndOf="parent"
+
+            app:layout_constraintStart_toEndOf="@+id/not_a_member"
+            app:layout_constraintTop_toTopOf="parent" />
+
+    </androidx.constraintlayout.widget.ConstraintLayout>
+
+
+
```
Comment (msarelo) [inline app/src/main/res/layout/fragment_forgot_password.xml:139]:
```suggestion
```

File: domain/build.gradle [hunk 8]
```diff
@@ -10,4 +10,5 @@ java {
 dependencies {
     implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.0")
     testImplementation 'junit:junit:4.13.2'
+
```
Comment (msarelo) [inline domain/build.gradle:13]:
```suggestion
```

File: domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.safetyheads.domain.entities.ResetPassword
+import com.safetyheads.domain.repositories.UserRepository
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class ResetPasswordUseCase(private val repository: UserRepository) :
+    BaseUseCase<ResetPasswordUseCase.ResetParam, ResetPassword> {
+
+    class ResetParam(val email: String) : BaseUseCase.Params
```
Comment (msarelo) [inline domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt:11]:
```suggestion
    value class ResetParam(val email: String) : BaseUseCase.Params
```

File: domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    override suspend fun invoke(parameter: ResetParam): Flow<Result<ResetPassword>> {
+        return flow {
+            try {
+                repository.resetPassword(parameter.email).collect { resetPasswordEntity ->
+                    if (resetPasswordEntity.isSuccess)
+                        emit(Result.success(resetPasswordEntity))
+                    else
+                        emit(Result.failure(resetPasswordEntity.error ?: Exception()))
```
Comment (msarelo) [inline domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt:20]:
```suggestion
                        emit(Result.failure(resetPasswordEntity.error ?: Exception("Reset password failed")))
```

File: domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+                }
+            } catch (error: java.lang.Exception) {
```
Comment (msarelo) [inline domain/src/main/java/com/safetyheads/domain/usecases/ResetPasswordUseCase.kt:22]:
```suggestion
            } catch (error: Exception) {
```

**Review by msarelo — CHANGES_REQUESTED** [review]

**Comment by Zalezny** [comment]
Snackbar with navigation and safeargs (it's only work on root activity):

https://user-images.githubusercontent.com/65240240/232505668-8bff7cf1-8ed7-45bb-b8d1-27be2609d743.mp4

**Review by msarelo — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/snackbar/View.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package com.safetyheads.akademiaandroida.forgotpasswordfragment.snackbar
+
+import android.view.View
+import android.view.ViewGroup
+import android.widget.FrameLayout
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+
+
+internal fun View?.findSuitableParent(): ViewGroup? {
```
[the remark this answers is not in the document]
Comment (Zalezny) [reply app/src/main/java/com/safetyheads/akademiaandroida/forgotpasswordfragment/snackbar/View.kt:9]: I'm not sure that is in android-ktx. I can't even find findSuitableParent() in official docs about androidx view. When I removed it from the project, nothing was found. It is just in Snackbar.java implemented:

image

I checked it on Github and they paste this code into their projects every time:

https://github.com/search?l=Kotlin&o=desc&q=findSuitableParent&s=indexed&type=Code

**Pull request merged**